include/ctype.h

```diff
@@ -37,21 +37,50 @@
 #ifndef __INCLUDE_CTYPE_H
 #define __INCLUDE_CTYPE_H
 
-/* There is no consistent ctype implementation, just a smattering of
- * functions.  Individually, they are okay, but a more standard, data lookup
- * approach would make more sense if used extensively.
- */
-
 /****************************************************************************
  * Included Files
  ****************************************************************************/
 
 #include <nuttx/compiler.h>
+#include <nuttx/arch.h>
 
 /****************************************************************************
  * Pre-processor Definitions
  ****************************************************************************/
 
+/* The classification flags */
+
+#define _U          0x01      /* upper */
+#define _L          0x02      /* lower */
+#define _N          0x04      /* digit */
+#define _S          0x08      /* space */
+#define _P          0x10      /* punct */
+#define _C          0x20      /* cntrl */
+#define _X          0x40      /* a-f|A-F */
+#define _B          0x80      /* blank */
+
+#define __ctype_lookup(c) up_romgetc(__ctype_ + (c))
```

Comment:
The AVR up_romgetc() is not accessible in user-space for PROTECTED and KERNEL builds.  It should should be!

Reply:
Yes, up_romget need move from arch/ to libc/machine.

Reply:
> 
> 
> Yes, up_romget need move from arch/ to libc/machine, we need another patch for that.

I don't think so.  I don't think that the up_romgetc() is need at all if the IOBJ (__flash) storage class is used.  Again, look at the  code example at https://gcc.gnu.org/onlinedocs/gcc/Named-Address-Spaces.html

    /* Locate array[] in flash memory */
    const __flash int array[] = { 3, 5, 7, 11, 13, 17, 19 };
    
    int i = 1;
    
    int main (void)
    {
       /* Return 17 by reading from flash memory */
       return array[array[i]];
    }

If the IOBJ (aka __flash) storage class is used, then the compiler will automatically generate the correct assembly language sequence to access the data in I-Space.  There is no need for up_romgetc() if that storage class is used.

Reply:
If so, we can directly remove up_romgetc.

Reply:
> 
> 
> If so, we can directly remove up_romgetc.

Yes, I think we can remove it.  I think we might break some AVR configurations that might not have the linker script setup to correctly place the objects.  I will need to look.